File: backend/app/api/v1/endpoints/export.py
"""
export.py

Endpoint for exporting the authenticated user's workout data.
"""

from datetime import datetime, timezone
from typing import Annotated, Literal

from fastapi import APIRouter, Depends, Query
from fastapi.responses import Response
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.services import export as export_service

router = APIRouter()


@router.get("/")
def export_user_data(
    a_db: Annotated[Session, Depends(get_db)],
    a_current_user: Annotated[User, Depends(get_current_user)],
    a_format: Annotated[Literal["json"], Query(alias="format")] = "json",
) -> Response:
    """Download a JSON export of all user-owned workout data."""
    payload = export_service.build_export_for_user(a_db, a_current_user)
    filename_date = datetime.now(timezone.utc).date().isoformat()
    filename = f'attachment; filename="progressive-overload-export-{filename_date}.{a_format}"'
    return Response(
        content=payload.model_dump_json(indent=2),
        media_type="application/json",
        headers={"Content-Disposition": filename},
    )

File: frontend/src/api/export.ts
/**
 * Export API: download a JSON attachment for the authenticated user.
 */
import { apiClient } from "./client";

const EXPORT_PREFIX = "/api/v1/export";

function getFilenameFromDisposition(contentDisposition?: string): string {
  const match = contentDisposition?.match(/filename="?(?<filename>[^"]+)"?/i);
  return match?.groups?.filename || "progressive-overload-export.json";
}

export async function downloadExportFile(format: "json" = "json"): Promise<string> {
  const response = await apiClient.get<Blob>(`${EXPORT_PREFIX}/`, {
    params: { format },
    responseType: "blob",
  });
  const filename = getFilenameFromDisposition(response.headers["content-disposition"]);
  const blob = new Blob([response.data], {
    type: response.headers["content-type"] || "application/json",
  });
  const objectUrl = URL.createObjectURL(blob);
  const link = document.createElement("a");

  link.href = objectUrl;
  link.download = filename;
  document.body.appendChild(link);
  link.click();
  link.remove();
  URL.revokeObjectURL(objectUrl);

  return filename;
}
